File: src/slang/parse/BinaryOpParselet.java
package slang.parse;

import slang.lex.Token;

public class BinaryOpParselet implements InfixParselet {
    private int precedence;
    private boolean isRight;

    BinaryOpParselet(int precedence, boolean isRight) {
        this.precedence = precedence;
        this.isRight = isRight;
    }

    @Override
    public Expr parse(Parser parser, Expr left, Token token) {
        // TODO(chris): Skip newlines here???
        Expr right = parser.expression(precedence - (isRight ? 1 : 0));
        return new Expr.Binary(left, token, right);
    }

    @Override
    public int getPrecedence() {
        return precedence;
    }
}

File: src/slang/parse/BinaryOpParselet.scala
package slang.parse

import slang.lex.Token

class BinaryOpParselet(val precedence: Int, val isRight: Boolean) extends InfixParselet {
    override def parse(parser: Parser, left: Expr, token: Token): Expr = {
        parser.skipNewlines
        val right = parser.expression(precedence - (if (isRight) 1 else 0))
        Expr.Binary(left, token, right);
    }

    override def getPrecedence: Int = precedence
}
